pyback/storage.py

```diff
@@ -85,26 +92,44 @@ def retrieve_file(self, checksum, dst_file_path, file_permissions):
         if not os.path.islink(dst_file_path):
             os.chmod(dst_file_path, file_permissions)
 
-    def store_checkpoint(self, checkpoint, checkpoint_id):
-        tree_file_path = os.path.join(self.tree_path, checkpoint_id + self.TREE_FILE_EXT)
+    def store_checkpoint(self, checkpoint):
+        tree_file_path = os.path.join(self.tree_path, checkpoint.meta.to_string() + self.TREE_FILE_EXT)
         if os.path.isfile(tree_file_path):
-            raise FileExistsError(checkpoint_id)
+            raise FileExistsError(checkpoint)
         if not os.path.exists(os.path.dirname(tree_file_path)):
             os.makedirs(os.path.dirname(tree_file_path))
         with open(tree_file_path, 'w') as f:
             f.write(checkpoint.to_json())
 
-    # TODO We should probably also retrieve some metadata about the checkpoint
-    #  except from the name to allow the user to chose between checkpoints depending on a timestamp, name etc.
-    def retrieve_checkpoint_ids(self):
-        return [checkpoint[:-len(self.TREE_FILE_EXT)] for checkpoint in os.listdir(self.tree_path) if checkpoint[-len(self.TREE_FILE_EXT):] == self.TREE_FILE_EXT]
+    def retrieve_checkpoint_metas(self):
+        return [CheckpointMeta.from_string(checkpoint_file[:-len(self.TREE_FILE_EXT)])
+                for checkpoint_file in os.listdir(self.tree_path)
+                if checkpoint_file[-len(self.TREE_FILE_EXT):] == self.TREE_FILE_EXT]
 
     def retrieve_checkpoint(self, checkpoint_id):
-        checkpoint_file_path = os.path.join(self.tree_path, checkpoint_id + self.TREE_FILE_EXT)
-        if not os.path.isfile(checkpoint_file_path):
-            raise FileNotFoundError(checkpoint_id)
-        with open(checkpoint_file_path, 'r') as f:
-            return Checkpoint.from_json(f.read())
+        def checkpoint_from_json_file(checkpoint_file):
+            with open(os.path.join(self.tree_path, checkpoint_file), 'r') as f:
+                return Checkpoint.from_json(f.read())
+
+        checkpoint_files = [c for c in os.listdir(self.tree_path)
+                            if c[-len(self.TREE_FILE_EXT):] == self.TREE_FILE_EXT]
+
+        # checkpoint meta
+        if isinstance(checkpoint_id, CheckpointMeta):
+            checkpoint_file = checkpoint_id.to_string() + self.TREE_FILE_EXT
+            if checkpoint_file in checkpoint_files:
+                return checkpoint_from_json_file(checkpoint_file)
+        elif isinstance(checkpoint_id, str):
+            # checksum
+            for checkpoint_file in checkpoint_files:
```

Comment:
If we match the first checkpoint file with a matching hash, we should ensure that the checkpoints are read in reversed time order.